lib/rex/proto/ldap/server.rb

```diff
@@ -0,0 +1,292 @@
+  # @param tcp [TrueClass, FalseClass] Listen on TCP socket
+  # @param ldif [String] LDIF data
+  # @param ctx [Hash] Framework context for sockets
+  # @param dblock [Proc] Handler for :dispatch_request flow control interception
+  # @param sblock [Proc] Handler for :send_response flow control interception
+  #
+  # @return [Rex::Proto::LDAP::Server] LDAP Server object
+  attr_reader :serve_udp, :serve_tcp, :sock_options, :udp_sock, :tcp_sock, :syntax, :ldif
+  def initialize(lhost = '0.0.0.0', lport = 389, udp = true, tcp = true, ldif = nil, comm = nil, ctx = {}, dblock = nil, sblock = nil)
+    @serve_udp    = udp
+    @serve_tcp    = tcp
+    @sock_options = {
+      'LocalHost' => lhost,
+      'LocalPort' => lport,
+      'Context'   => ctx,
+      'Comm'      => comm
+    }
+    @ldif         = ldif
+    self.listener_thread = nil
+    self.dispatch_request_proc = dblock
+    self.send_response_proc = sblock
+  end
+
+  #
+  # Check if server is running
+  #
+  def running?
+    self.listener_thread and self.listener_thread.alive?
+  end
+
+  #
+  # Start the LDAP server
+  #
+  def start
+
+    if self.serve_udp
+      @udp_sock = Rex::Socket::Udp.create(self.sock_options)
+      self.listener_thread = Rex::ThreadFactory.spawn("UDPLDAPServerListener", false) {
+        monitor_listener
+      }
+    end
+
+    if self.serve_tcp
+      @tcp_sock = Rex::Socket::TcpServer.create(self.sock_options)
+      self.tcp_sock.on_client_connect_proc = Proc.new { |cli|
+        on_client_connect(cli)
+      }
+      self.tcp_sock.on_client_data_proc = Proc.new { |cli|
+        on_client_data(cli)
+      }
+      self.tcp_sock.start
+      if !self.serve_udp
+        self.listener_thread = tcp_sock.listener_thread
+      end
+    end
+  end
+
+  #
+  # Stop the LDAP server
+  #
+  def stop
+    ensure_close = [self.udp_sock, self.tcp_sock].compact
+    begin
+      self.listener_thread.kill if self.listener_thread.respond_to?(:kill)
+      self.listener_thread = nil
+    ensure
+      while csock = ensure_close.shift
+        csock.stop if csock.respond_to?(:stop)
+        csock.close unless csock.respond_to?(:close) and csock.closed?
+      end
+    end
+  end
+
+  #
+  # Process client request, handled with dispatch_request_proc if set
+  #
+  # @param cli [Rex::Socket::Tcp, Rex::Socket::Udp] Client sending the request
+  # @param data [String] raw LDAP request data
+  def dispatch_request(cli, data)
+    if self.dispatch_request_proc
+      self.dispatch_request_proc.call(cli,data)
+    else
+      default_dispatch_request(cli,data)
+    end
+  end
+
+  #
+  # Default LDAP request dispatcher
+  #
+  # @param cli [Rex::Socket::Tcp, Rex::Socket::Udp] Client sending the request
+  # @param data [String] raw LDAP request data
+  def default_dispatch_request(cli, data)
+    return if data.strip.empty?
+    data.extend(Net::BER::Extensions::String)
+    while pdu = Net::LDAP::PDU.new(data.read_ber!(Net::LDAP::AsnSyntax))
```

Comment:
So it looks like this is intending to handle an encoded stream of requests. It'd be really nice if each PDU was handled individually as an option when using the library. Something like an `on_request_pdu` callback that would handle exactly one PDU request that could then dispatch to the services default handler. That would enable modules to catch the request types they want to handle themselves and forward the rest to the default without needing all the error handling and looping logic.

Reply:
Also I was encountering an exception here when the stream was depleted.

```suggestion
    while !data.empty? && pdu = Net::LDAP::PDU.new(data.read_ber!(Net::LDAP::AsnSyntax))
```

Reply:
For now i am logging these, but yeah, we should revise as suggested (adding as a TODO)